src/Nethermind/Nethermind.State/VerkleTransition/AccountTreeMigrator.cs

```diff
@@ -0,0 +1,229 @@
+        {
+            currentPath.Add(Bytes.FromHexString(trieVisitContext.BranchChildIndex.Value.ToString("x2")));
+        }
+    }
+
+    public void VisitExtension(TrieNode node, TrieVisitContext trieVisitContext)
+    {
+        List<byte[]> currentPath = trieVisitContext.IsStorage ? _currentPathStorage : _currentPath;
+        if (node.Key is not null)
+        {
+            currentPath.Add(node.Key);
+        }
+        else if (currentPath.Count > 0)
+        {
+            // Pop off last value when exiting an extension node
+            currentPath.RemoveAt(currentPath.Count - 1);
+        }
+    }
+
+    private readonly AccountDecoder decoder = new();
+
+    /// <summary>
+    /// Converts a 0-padded hex string to a byte array, i.e. from '050502' to '552'.
+    /// TODO: check if this is the best way to do this
+    /// </summary>
+    /// <param name="paddedHex"></param>
+    /// <returns></returns>
+    private static byte[] ConvertPaddedHexToBytes(string paddedHex)
+    {
+        // Remove the padding (leading zeros)
+        string unpadded = string.Concat(
+            paddedHex
+                .Where((c, i) => i % 2 == 1)
+                .Select(c => c.ToString())
+        );
+
+        // Convert the unpadded hex string to byte array
+        return Enumerable.Range(0, unpadded.Length / 2)
+            .Select(x => Convert.ToByte(unpadded.Substring(x * 2, 2), 16))
+            .ToArray();
+    }
+
+    private static byte[] ConstructFullHash(int branchIndex, byte[] nodeKey, List<byte[]> currentPath)
+    {
+        string branchPath = "";
+        if (!currentPath.IsNullOrEmpty())
+        {
+            branchPath = currentPath.ToArray().CombineBytes().ToHexString();
+        }
+        string branchIndexHex = branchIndex.ToString("x2");
+        string currentKey = nodeKey.ToHexString();
+        string addressHash = branchPath + branchIndexHex + currentKey;
+        string unpaddedAddressHash = ConvertPaddedHexToBytes(addressHash).ToHexString();
+
+        return Bytes.FromHexString(unpaddedAddressHash);
+    }
+
+    private static byte[] ConstructFullHash(byte[] nodeKey, List<byte[]> currentPath)
+    {
+        string branchPath = "";
+        if (!currentPath.IsNullOrEmpty())
+        {
+            branchPath = currentPath.ToArray().CombineBytes().ToHexString();
+        }
+        string currentKey = nodeKey.ToHexString();
+        string addressHash = branchPath + currentKey;
+        string unpaddedAddressHash = ConvertPaddedHexToBytes(addressHash).ToHexString();
+
+        return Bytes.FromHexString(unpaddedAddressHash);
+    }
+
+    public void VisitLeaf(TrieNode node, TrieVisitContext trieVisitContext, ReadOnlySpan<byte> value)
+    {
+        List<byte[]> currentPath = trieVisitContext.IsStorage ? _currentPathStorage : _currentPath;
+
+        while (currentPath.Count >= trieVisitContext.Level && trieVisitContext.Level > 0)
+        {
+            currentPath.RemoveAt(currentPath.Count - 1);
+        }
```

Comment:
i cant see this being used anywhere

Reply:
Currently still facing an edge case for large trees, where some of the leaves do not generate the full keccak hash based on my path tracking logic. Will need to spend more time debugging this